nh.c

```diff
@@ -64,6 +64,56 @@ __attribute__((visibility("hidden"))) void nh_failsafe_destroy(nh_failsafe_t *fs
 // used afterwards.
 __attribute__((visibility("hidden"))) void nh_failsafe_uninstall(nh_failsafe_t *fs);
 
+// The following blacklists help prevent any whoopsies when deleting files during uninstall
+
+// These files are important for recovery purposes.
+static const char *delete_file_blacklist[] = {
+    "/bin/sh",
+    "/etc/init.d/rcS",
+    "/usr/local/Kobo/pickel",
+};
+
+// Let's not allow deleting any files in /bin, /sbin, or u-boot stuff.
+static const char *delete_prefix_blacklist[] = {
+    "/bin",
+    "/sbin",
+    "/etc/u-boot",
+};
+
+// This is for the schmuck who forgot to terminate his array of paths
+// and accidentially deleted '/bin/sh' due to an out-of-bounds read.
+static void nh_file_delete(const char *path) {
+    if (path) {
+        for (unsigned int i = 0; i < (sizeof(delete_file_blacklist) / sizeof(*delete_file_blacklist)); i++) {
+            if (!strcmp(delete_file_blacklist[i], path)) {
+                nh_log("(NickelHook) not deleting blacklisted file %s", path);
+                return;
+            }
+        }
+        for (unsigned int i = 0; i < (sizeof(delete_prefix_blacklist) / sizeof(*delete_prefix_blacklist)); i++) {
+            if (!strncmp(delete_prefix_blacklist[i], path, strlen(delete_prefix_blacklist[i]))) {
+                nh_log("(NickelHook) not deleting %s with blacklisted prefix %s", path, delete_prefix_blacklist[i]);
+                return;
+            }
+        }
+        // remove(), to support both files and directories
+        if (remove(path) == -1) {
+            nh_log("(NickelHook) failed to delete %s, with error: %s", path, strerror(errno));
```

Comment:
Not that we care here, because we know we're going to build w/ NickelTC, hence glibc, but prefer the GNU extension `%m` printf token to `strerror`. It's both more portable (musl) and actually thread-safe (on both glibc & musl, even if the doc is unclear on that point, the code isn't) ;).

Reply:
Actually, I say that know, but I realize NM/NH itself is already using `strerror`, pinging @pgaskin  ;p.

Reply:
Oops, I keep forgetting that that format specifier exists ... I was meaning to use it, but it just is one of those things which always slips your mind when you actually get around to it.